src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -471,19 +489,114 @@ void enforceMemberNotNull(SyntaxNode? syntaxOpt, LocalState state)
+                            checkMemberStateOnConstructorExit(method, member, state, thisSlot, exitLocation);
+                        }
+                    }
+                    else
+                    {
+                        do
+                        {
+                            foreach (var memberName in method.NotNullMembers)
+                            {
+                                enforceMemberNotNullOnMember(syntaxOpt, state, method, memberName);
+                            }
 
-                        method = method.OverriddenMethod;
+                            method = method.OverriddenMethod;
+                        }
+                        while (method != null);
                     }
-                    while (method != null);
                 }
             }
 
+            void checkMemberStateOnConstructorExit(MethodSymbol constructor, Symbol member, LocalState state, int thisSlot, Location? exitLocation)
+            {
+                var isStatic = !constructor.RequiresInstanceReceiver();
+                if (member.IsStatic != isStatic)
+                {
+                    return;
+                }
+
+                // This is not required for correctness, but in the case where the member has
+                // an initializer, we know we've assigned to the member and
+                // have given any applicable warnings about a bad value going in.
+                // Therefore we skip this check when the member has an initializer to reduce noise.
+                if (HasInitializer(member))
+                {
+                    return;
+                }
+                TypeWithAnnotations fieldType;
+                FieldSymbol? field;
+                switch (member)
+                {
+                    case FieldSymbol f:
+                        fieldType = f.TypeWithAnnotations;
+                        field = f;
+                        break;
+                    case EventSymbol e:
+                        fieldType = e.TypeWithAnnotations;
+                        field = e.AssociatedField;
+                        if (field is null)
+                        {
+                            return;
+                        }
+                        break;
+                    default:
+                        return;
+                }
+                if (field.IsConst)
+                {
+                    return;
+                }
+                if (fieldType.Type.IsValueType || fieldType.Type.IsErrorType())
+                {
+                    return;
+                }
+                if (!fieldType.NullableAnnotation.IsNotAnnotated())
+                {
+                    return;
+                }
+                var symbol = member switch
+                {
+                    FieldSymbol { AssociatedSymbol: PropertySymbol p } => p,
+                    _ => member
+                };
+                var annotations = symbol.GetFlowAnalysisAnnotations();
```

Comment:
Under the hood this is doing yet another `switch` on the types. This one may be cleaner to keep in but want to make sure you considered it.

Reply:
Yeah, I think it's best to keep it in this case.